design/one-pager-consuming-k8s-clusters.md

```diff
@@ -0,0 +1,321 @@
+it is helpful to separate the notions of a **Kubernetes cluster as a managed
+service** and a **Kubernetes cluster as a target for deploying workloads**. With
+this distinction made, we can assert that this proposal in no way modifies the
+managed service notion of a Kubernetes cluster, but instead formalizes this
+concept of a Kubernetes cluster as a deployment target.
+
+## Goals
+
+Because it is unlikely that Crossplane users will always want a 1-to-1
+namespace-to-cluster consumption model, it is desirable to support deploying
+workloads in any namespace to any Kubernetes cluster. However, it is also
+desirable to continue to treat managed Kubernetes services as a managed resource
+that is supplied by cloud providers. For this reason, any solution should be
+additive to the existing model.
+
+As a secondary goal for this design, it should be possible to use existing
+Kubernetes clusters (wherever they may be running) that were not provisioned by
+Crossplane as targets for deploying workloads. This concept has frequently been
+referred to as a "Bring Your Own Cluster" scenario.
+
+## Proposal
+
+This proposal can be separated into three main categories:
+1. The implementation of a new `KubernetesTarget` custom resource and
+   corresponding controllers.
+1. The modification of the scheduling behavior for `KubernetesApplication`.
+1. The implementation of a controller that automatically creates a
+   `KubernetesTarget` resource when a `KubernetesClaim` is bound to a Kubernetes
+   cluster managed resource. 
+
+### KubernetesTarget
+
+The purpose of the `KubernetesTarget` resource is to effectively "publish"
+Kubernetes clusters for usage in a namespace. It is namespace-scoped and can
+only reference a cluster-scoped Kubernetes cluster managed resource in its
+`clusterRef` field, which will result in the setting of its
+`connectionSecretRef` field, or a local namespace-scoped `Secret` directly in
+the `connectionSecretRef` field. If both the `clusterRef` field and
+`connectionSecretRef` field are populated at time of creation, the
+`connectionSecretRef` field will take precedence. The creation of a
+`KubernetesTarget` resources will be watched by controllers in stacks that
+provide Kubernetes cluster managed resources (i.e. a specific controller must be
+implemented in each stack that wishes to provide a Kubernetes cluster managed
+resource that `KubernetesApplication` resources can be scheduled to). These
+controllers will check to see if the `KubernetesTarget` references their cluster
+type n its `clusterRef` and, if so, will propagate the cluster's connection
+`Secret` to the namespace of the `KubernetesTarget`. A local object reference to
+the `Secret` will then be set on the `KubernetesTarget` object in the
+`connectionSecretRef` field. If a `KubernetesTarget` is created with a reference
+to a local `Secret`, it will be ignored by these controllers as the required
+`Secret` must already be present in the namespace. An example of a
```

Comment:
I think it should keep refreshing that secret as credentials may change, for example EKS clusters. Maybe a logic like `if connectionSecretRef empty, fill with a name. If it's filled go ahead and populate that secret` could work instead of checking the reference's existence.

Reply:
This is a good point, and as you mentioned with EKS we will actually *have* to do this. This gets a little bit complicated as the current proposal says that we don't do anything if `connectionSecretRef` is populated, but what we actually want to know is if the `connectionSecretRef` is populated by a manually created `Secret`. I think the easiest fix for this problem would be:

1. Controllers reconcile any `KubernetesTarget` that references their managed resource type in the `clusterRef`.
1. If the `connectionSecretRef` is unset then the `Secret` is propagated to the namespace and the reference is set.
1. If the `connectionSecretRef` is set, then it is updated to match the managed resource's connection `Secret`.

The primary areas of concern here to me are:
1. If you create a `KubernetesTarget` that with a `connectionSecretRef` to a manually created "BYO Cluster" `Secret` *and* a `clusterRef`, that manually created `Secret` will be overwritten by the `clusterRef` connection `Secret`. This seems fine to me. Instead of saying that we defer to the `connectionSecretRef` when it is set, we are saying that if a `clusterRef` is set then we will always use it.
1. Because these controllers are separate from the managed resource controllers (where continuous secret propagation currently happens), there is a longer lag time from when connection information changes to the `Secret` being updated. I also, don't see this as a huge issue when framed as a max of a 2 minute lag (managed reconciler long wait + KubernetesTarget reconciler long wait) compared to a max of a 1 minute lag (managed reconciler long wait) in the current system where the managed reconciler continuously published connection secret updates.

@muvaf I think all of this is in-line with your thoughts, but let me know if I have deviated. Thanks for bringing this up :)

@negz could you indicate if this update sounds right to you?

Reply:
@negz pointed out in stand-up today that the existing [secret reconciler](https://github.com/crossplaneio/crossplane-runtime/blob/master/pkg/resource/secret_reconciler.go) can be used to accomplish this. If I understand correctly, with EKS (where this reconciler is already [used](https://github.com/crossplaneio/stack-aws/blob/master/pkg/controller/compute/secret.go)), the controller could just be modified to also watch secrets controlled by a `KubernetesTarget`.

```go
func (c *EKSClusterSecretController) SetupWithManager(mgr ctrl.Manager) error {
	p := resource.NewPredicates(resource.AnyOf(
		resource.AllOf(resource.IsControlledByKind(v1alpha1.KubernetesClusterGroupVersionKind), resource.IsPropagated()),
		resource.AllOf(resource.IsControlledByKind(v1alpha1.KubernetesTargetGroupVersionKind), resource.IsPropagated()),
		resource.AllOf(resource.IsControlledByKind(v1alpha3.EKSClusterGroupVersionKind), resource.IsPropagator()),
	))

	return ctrl.NewControllerManagedBy(mgr).
		Named(strings.ToLower(fmt.Sprintf("connectionsecret.%s.%s", v1alpha3.EKSClusterKind, v1alpha3.Group))).
		Watches(&source.Kind{Type: &corev1.Secret{}}, &resource.EnqueueRequestForPropagator{}).
		For(&corev1.Secret{}).
		WithEventFilter(p).
		Complete(resource.NewSecretPropagatingReconciler(mgr))
}
```
And we would only need to set the [required annotations](https://github.com/crossplaneio/crossplane-runtime/blob/bd43e78aa7637acd75c371104bac2cc4778f6e30/pkg/resource/secret_reconciler.go#L41) when the `Secret` is initially propagated. This would allow the `KubernetesTarget` controller to function as currently proposed in this design and refuse to operate on `KubernetesTarget` resources where the `connectionSecretRef` is already set.

Reply:
One thing we'd have to consider with the above controller is that a source secret can currently only be propagated to one destination, due to how the annotations work. Currently we only instantiate the secret propagation controller for EKS, but we set the propagation annotations on connection secrets for all managed resources at claim binding time. We'll need to update the annotations to support propagating a secret to multiple destinations.

Another thing to keep in mind is that the secret propagation reconciler will not, by design, propagate _to_ a secret that does consent to the process via a set of annotations. This means we won't be able to (and IMO should not try to) let `KubernetesTarget` specify an existing secret to propagate to.

So I believe the behaviour should be, if the user specifies:

* Only `connectionSecretRef` no propagation happens, we just try to use the referenced secret.
* Only `clusterRef` we'll create and start propagating to a secret.
* Both, and the referenced secret does not exist, we'll create it and start propagating to it.
* Both, and the referenced secret exists without the consenting propagation annotations we'll refuse to propagate to the existing secret.